build_retrieval_dataset.py

```diff
@@ -0,0 +1,389 @@
+    """
+    Returns a random coin.
+    :param countries: the list of countries to choose from
+    :param cents: the list of coin values to choose from, in cents
+    :param data_path: the folder containing the dataset
+    :return: coin value, coin image
+    """
+
+    coin_names = {
+        200: '2',
+        100: '1',
+        50: '50ct',
+        20: '20ct',
+        10: '10ct',
+        5: '5ct',
+        2: '2ct',
+        1: '1ct'
+    }
+    country = random.choice(countries)
+    coin_value = random.choice(cents)
+    coin_name = coin_names[coin_value]
+    path = os.path.join(data_path, country, coin_name)
+
+    coin_files = [os.path.join(path, f) for f in os.listdir(path) if os.path.isfile(os.path.join(path, f))]
+    coin_file = random.choice(coin_files)
+
+    return coin_value, cv.imread(coin_file)
+
+
+def create_bgr_image(height, width, bg=(0,0,0)):
+    """
+    Creates a image with the given height, width, and background.
+    :param height: height of the picture
+    :param width: width of the picture
+    :param bg: background color, given as (b,g,r) tuple
+    :return: an image of size height * width with bg as background color
+    """
+    img = np.zeros((height, width, 3), np.uint8)
+    img[:, :] = bg
+    return img
+
+
+def rotate_pic(inp_pic, phi):
+    """
+    Rotates an input picture counter-clockwise around its center with an given angle.
+    :param inp_pic: input picture
+    :param phi: rotation angle
+    :return: output picture
+    """
+    y, x = inp_pic.shape[:2]
+    matrix = cv.getRotationMatrix2D((x / 2, y / 2), phi, 1.0)
+    out_pic = cv.warpAffine(inp_pic, matrix, (x, y))
+    return out_pic
+
+
+def hough_circle_detection(inp_pic, blur_strgth, hough_dp=1, minRadius=180, maxRadius=190):
+    """
+    Detects a circle (through Hough Transformation) and returns the coordinates (x_ctr, y_ctr, r)
+    :param inp_pic: input picture
+    :param blur_strgth: choose between "low" and "high" blurring
+    :param hough_dp: Inverse ratio of the accumulator resolution to the image resolution.
+    For example, if dp=1 , the accumulator has the same resolution as the input image.
+    If dp=2 , the accumulator has half as big width and height.
+    :param minRadius: minimum circle radius
+    :param maxRadius: maximum circle radius
+    :return: coordinates (x_ctr, y_ctr) and the circle radius r of the found circle
+    """
+    inp_pic_grey = cv.cvtColor(inp_pic, cv.COLOR_BGR2GRAY)
+    if blur_strgth == "low":
+        inp_pic_grey_blurred = cv.GaussianBlur(inp_pic_grey, (3,3), 1, 1)
+    elif blur_strgth == "high":
+        inp_pic_grey_blurred = cv.GaussianBlur(inp_pic_grey, (9,9), 2, 2)
+    #cv.imshow("grey_blurred", inp_pic_grey_blurred)
+    # HoughCircles(image, method, dp, minDist, circles=None, param1=None, param2=None, minRadius=None, maxRadius=None)
+    circles = cv.HoughCircles(inp_pic_grey_blurred, cv.HOUGH_GRADIENT, hough_dp, 20, minRadius, maxRadius) #minDist = 20
```

Comment:
Here parameters are passed to the HoughCircles function in an ambiguous way, I think minRadius and maxRadius are actually passed as param1 and param2. This last two parameters also are very important and should be set in a convenient way. It's better to do:
`
circles = cv.HoughCircles(inp_pic_grey_blurred, cv.HOUGH_GRADIENT, hough_dp, 20, minRadius=minRadius, maxRadius=maxRadius)
`

Reply:
parameter setting adapted to the following:
`circles = cv.HoughCircles(inp_pic_grey_blurred, cv.HOUGH_GRADIENT, hough_dp, circles=1, minDist=20, minRadius=minRadius, maxRadius=maxRadius)`

(comment: if circles=None no circles found, unsure why..)